### PR TITLE
8 implement edge and node classification network

### PR DESCRIPTION
**Pull request by chris-soelistyo**
Still in a sort-of conceptual stage.

What's been done
- Changed `GCN` to output `x, embedding ` rather than just `x`.
- Adapted the `torch_geometric.data.Data` files to also include `edge_label`
- Added an `edge_criterion` loss to `train.py`

Questions
- Should `edge_criterion` be made to be a subclass of `_WeightedLoss`, or is it alright to keep it as a bare function?

What's not been done (yet)
- Changing `.test()` method to output edge classification results
- Tests (will probably do this after #64 has been merged)

### CODE REVIEW

File: grace/models/classifier.py [hunk 1]
```diff
@@ -7,48 +7,36 @@
...
     def __init__(
         self,
+        input_channels: int,
+        hidden_channels: int,
         *,
-        input_dims: int = 3,
-        embedding_dims: int = 64,
-        output_dims: int = 2,
-    ) -> None:
+        output_classes: int = 2,
```
Comment (quantumjot) [inline grace/models/classifier.py:17]: Is there a reason you removed all of this?

File: grace/models/train.py [hunk 2]
```diff
@@ -29,26 +33,45 @@ def train_model(
...
         for data in train_loader:
-            out = model(data.x, data.edge_index, data.batch)
-            loss = criterion(out, data.y)
+            out_x, out_embedding = model(data.x, data.edge_index, data.batch)
+            loss_node = node_criterion(out_x, data.y)
+            loss_edge = edge_criterion(
+                out_embedding, data.edge_label, data.edge_index
+            )
+
+            loss = loss_node + loss_edge
```
Comment (quantumjot) [inline grace/models/train.py:57]: Nice!

File: grace/models/classifier.py (hunk 1 above)
Comment (quantumjot) [reply grace/models/classifier.py:17]: It was added in https://github.com/alan-turing-institute/grace/pull/72

**Comment by quantumjot** [comment]
@chris-soelistyo - is this still a work in progress or do you want reviews? If the former, I'd convert to a draft PR until you're ready

File: grace/models/classifier.py (hunk 1 above)
Comment (chris-soelistyo) [reply grace/models/classifier.py:17]: Oh!! sorry will change back to it now

**Comment by chris-soelistyo** [comment]
@quantumjot Cool, I'll change to a draft PR then. Was mostly seeking comments re: whether this was the right scheme/structure, but if so, I'll draft it for now and get working on all the stuff in the "not been done" list.

File: grace/models/classifier.py (hunk 1 above)
Comment (chris-soelistyo) [reply grace/models/classifier.py:17]: So just to be clear; `input_dims` etc is basically the same as `input_channels` (i.e., an integer) and not a full `Tuple` specification of dims?

Comment (quantumjot) [reply grace/models/classifier.py]: Yeah - although maybe we should keep channels or features rather than dims?

Comment (quantumjot) [reply grace/models/classifier.py]: torch and torch geometric both use different terminology

Comment (chris-soelistyo) [reply grace/models/classifier.py]: Ah ok; I think channels probably works better? Dims to me suggests a full shape, but let me know if you disagree!

Comment (chris-soelistyo) [reply grace/models/classifier.py]: Or even "dim" as a singular works better imo

Comment (quantumjot) [reply grace/models/classifier.py]: let's revert to channels, but keep the docstrings (updated obv.)

Comment (chris-soelistyo) [reply grace/models/classifier.py]: @quantumjot ok! and `output_classes` ?

Comment (mooniean) [reply grace/models/classifier.py]: this is the number of features coming from the extractor, right? so definitely needs to be documented properly!

File: grace/models/classifier.py [hunk 3]
```diff
@@ -11,44 +11,44 @@ class GCN(torch.nn.Module):
...
     ----------
-    input_dims : int
-        The dimensions of the input.
-    embedding_dims : int
-        The dimensions of the hidden embeddings.
+    input_channels : int
+        The dimension of the input; i.e., length of node feature vectors
+    embedding_channels : int
+        The dimension of the hidden embeddings.
     output_dims : int
```
Comment (quantumjot) [inline grace/models/classifier.py:18]:
```suggestion
    output_classes : int
```

File: grace/models/train.py [hunk 4]
```diff
@@ -4,13 +4,32 @@
...
+import torch.nn.functional as F
+
+from grace.base import Annotation
+
+
+def edge_criterion(
+    embed: torch.Tensor,
+    target: torch.Tensor,
+    edge_index: torch.Tensor,
+    masked_class: int,
```
Comment (quantumjot) [inline grace/models/train.py:16]:
```suggestion
    masked_class: Annotation,
```

File: grace/models/train.py [hunk 5]
```diff
@@ -4,13 +4,32 @@
...
+    return F.cross_entropy(edge_score[mask], target[mask])
+
 
 def train_model(
     model: torch.nn.Module,
     dataset: List[torch_geometric.data.Data],
     *,
     epochs: int = 100,
     batch_size: int = 64,
+    masked_class: int = Annotation.UNKNOWN,
```
Comment (quantumjot) [inline grace/models/train.py:32]:
```suggestion
    masked_class: Annotation = Annotation.UNKNOWN,
```

File: grace/models/train.py [hunk 6]
```diff
@@ -29,33 +48,62 @@ def train_model(
         # weight_decay=5e-4),
     )
 
-    criterion = torch.nn.CrossEntropyLoss()
+    node_criterion = torch.nn.CrossEntropyLoss(
+        ignore_index=Annotation.UNKNOWN, size_average=True
+    )
```
Comment (quantumjot) [inline grace/models/train.py:53]: I'm not sure what happens if the whole subgraph is unknown?

File: grace/models/train.py [hunk 7]
```diff
@@ -29,33 +48,62 @@ def train_model(
...
+            pred = out_x.argmax(
+                dim=1
+            )  # Use the class with highest probability.
+            correct_nodes += int((pred == data.y).sum())
+
+            # edge
+            src, dst = data.edge_index
+            edge_score = (out_embedding[src] * out_embedding[dst]).sum(dim=-1)
+            num_edges += edge_score.size(dim=0)
+            correct_edges += int((torch.round(pred) == data.edge_label).sum())
```
Comment (quantumjot) [inline grace/models/train.py:94]: What is the function of this `round`? Have to admit I'm a little confused by this, because it seems like this is for a binary classification - whereas, we may have multiple edge classes 😕

File: grace/models/train.py [hunk 8]
```diff
@@ -29,33 +48,62 @@ def train_model(
...
         print(
-            f"Epoch: {epoch:03d}, Train Acc: {train_acc:.4f},"
-            f" Test Acc: {test_acc:.4f}"
+            f"Epoch: {epoch:03d}, Loss: {loss:.4f},"
+            f" Node Loss: {loss_node:.4f}, Edge Loss: {loss_edge:.4f},"
+            f" Train Acc (Node): {train_acc_node:.4f},"
+            f" Train Acc (Edge): {train_acc_edge:.4f},"
+            f" Test Acc (Node): {test_acc_node:.4f},"
+            f" Test Acc (Edge): {test_acc_edge:.4f}"
         )
```
Comment (quantumjot) [inline grace/models/train.py:109]: At some point we should add some proper logging. But this is fine for now

File: grace/models/train.py (hunk 7 above)
Comment (quantumjot) [reply grace/models/train.py:94]: Is this binary because you're ignoring the `UNKNOWN` class?

Comment (quantumjot) [reply grace/models/train.py:94]: I see, this works because TP and TN are 1 and 0 in the `Annotation` enum and the unknown is ignored?

**Review by quantumjot — COMMENTED** [review]
Some questions and suggestions for you.

File: grace/models/train.py (hunk 6 above)
Comment (chris-soelistyo) [reply grace/models/train.py:53]: @quantumjot In that case the node loss reduces to NaN... same behaviour with the `edge_criterion` custom loss; would it be worth adding `loss_node = torch.nan_to_num(loss_node)` etc?

File: grace/models/train.py (hunk 7 above)
Comment (chris-soelistyo) [reply grace/models/train.py:94]: Yup! That's why it works.
Though now that you've mentioned it, it doesn't work for if we add more edge classes than just TP and TN

Comment (quantumjot) [reply grace/models/train.py:94]: OK - I think that's fine for now. I would just open an issue to make sure that we don't forget this.

File: grace/models/train.py (hunk 6 above)
Comment (chris-soelistyo) [reply grace/models/train.py:53]: Ah, like some if statement, e.g.

`if torch.all(torch.where(data.edge_label == Annotation.UNKNOWN, True, False)):
    continue`

?

File: grace/models/train.py (hunk 7 above)
Comment (chris-soelistyo) [reply grace/models/train.py:94]: @quantumjot Do you reckon this requires modification for now?

Comment (quantumjot) [reply grace/models/train.py:94]: Leave it for now, but open an issue once this is merged

File: grace/models/train.py (hunk 6 above)
Comment (chris-soelistyo) [reply grace/models/train.py:53]:
```
edge_label = [
          edge[GraphAttrs.EDGE_GROUND_TRUTH]
          for _, _, edge in sub_graph.edges(data=True)
      ]

if all([e == Annotation.UNKNOWN for e in edge_label]):
   continue
```
Would this work?

Comment (chris-soelistyo) [reply grace/models/train.py:53]: @quantumjot in `datasets.py`, that is

Comment (quantumjot) [reply grace/models/train.py:53]: Yeah - I think that works. It may also need to have this if all the nodes are empty.

Comment (quantumjot) [reply grace/models/train.py:53]: No, that the nodes (i.e. `GraphAttrs.NODE_GROUND_TRUTH]`) are also all `UNKNOWN`.

Comment (chris-soelistyo) [reply grace/models/train.py:53]: Cool, will add that as well
